src/utils/password.ts

```diff
@@ -0,0 +1,21 @@
+type PasswordRules = 'must contain 1 digit' | 'must contain 1 letter' | 'must have at least six characters';
```

Comment:
Apesar de o typescript ajudar a garantir com a tipagem, acho que pra criar um tipo union type desse seria mais interresante ter opções mais sucintas. Algo como:

```suggestion
type PasswordRules = 'contains_digit' | 'contains_letter' | 'min_size';
```

Outro ponto em relação a colocar o "six" dentro da string é que a gente pode querer alterar esse tamanho mínimo, então é bom que essa descrição fique mais genérica para que não precisemos mudar aqui tbm.

Reply:
Eu acabei usando mensagens mais descritivas por que eu acabo aproveitando elas na hora de gerar os erros. Aí todas vez que tiver que atualizar alguma regra, só precisa atualizar esse trecho do código:

```typescript
const passwordRules: Record<PasswordRules, PasswordRuleChecker> = {
  'must contain 1 digit': (password) => /\d+/.test(password),
  'must contain 1 letter': (password) => /[a-zA-Z]+/.test(password),
  'must have at least six characters': (password) => /.{6,}/.test(password),
};
```

A vantagem dessa abordagem também é que eu consigo saber exatemente quais as regras que foram violadas.

Reply:
você poderia ter outro mapa entre o erro (em opções mais sucintas como o alan apresentou) e as mensagens que voltaria para o usuario
com isso, a mensagem não fica fortemente atrelada ao type PasswordRules e vc não precisa se preocupar com breaking changes (imagina se vc guardasse os erros no banco para auditoria e fosse necessario trocar a mensagem, o caos que ia ser)

Reply:
Esse é o tipo de coisa que não vale muito a pena reaproveitar, porque são conceitos diferentes. Um deles é tipo do erro, que seu código usa pra identificar por meio de uma condição e o outro é a mensagem que será retornada. Como o Matheus citou um exemplo, acoplar essas duas coisas pode gerar dificuldades de manutenção no futuro. Faz sentido?

Reply:
Ahhh acho que saquei! Então o tipo `PasswordRules` deveria ser tipo uma tag para cada erro? E aí como o Matheus recomendou é melhor mapear a mensagem dos erros em outro lugar.

Reply:
Eu deixaria essas strings como constantes para facilitar a referência e não precisar digitar a mesma coisa toda a vez que for utilizar